active/0051-strategies.md

```diff
@@ -0,0 +1,368 @@
+}
+```
+
+`inputTree` will have the following structure:
+
+```ruby
+bundler:js:input/
+├── addon-modules/
+├── the-app-name-folder/
+├── node_modules/
+└── vendor/
+```
+
+where:
+
++ `addon-modules` is a folder that contains dependencies from Ember add-ons
++ `the-app-name-folder` is a folder that contains Ember application code
++ `node_modules` is a folder that contains node dependencies
++ `vendor` is a folder that contains other dependencies.
+
+Note, that  `toTree` method must return a broccoli tree.
+
+Strategies get access to the same input tree, they are not chained off of each
+other and they don't run in parallel.
+
+### Concatenation Strategy
+
+It is used by Ember CLI privately in the latest `canary` version to produce
+`assets/app-name.js` and `assets/vendor.js`. It takes a broccoli tree and
+returns a new concatenated broccoli tree.
+
+```javascript
+const concat = require('broccoli-concat');
+
+class ConcatenationStrategy {
+  constructor(options) { this.options = options; }
+
+  toTree(assembler, inputTree) {
+    return concat(inputTree, this.options);
+  }
+}
+```
+
+Where `this.options` mimics [`broccoli-concat`
+options](https://github.com/broccolijs/broccoli-concat#advanced-usage).
+
+### Example of possible strategies
+
+Below are some of the strategies that could be implemented.
+
+#### Assets Split Strategy
+
+One of the techniques for improving site speed is isolating changes throughout
+application deployments. Assuming  the application assets are uploaded to CDN,
+the reasoning is very simple: if `ember.js` or `jQuery` (possibly along with
+other third party libraries) don't change with every deployment, why bust CDN
+cache for them?
+
+This strategy could be provided out of the box by Ember CLI as a part of site
+speed improvement story.
+
+```javascript
+class AssetsSplitStrategy {
+  constructor(options) {
+    this.options = options;
+  }
+
+  toTree(assembler, inputTree) {
+    return concat(inputTree, {
+      headerFiles: [
+        jQueryPath, emberPath
+      ],
+      outputFile: 'vendor-static.js'
+    });
+  }
+}
+```
+
+#### Debug/Analyse Strategy
+
+Yet another application of strategies would be to run different analysis on the
+Ember applications. Take
+[broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser),
+for example. This could be a strategy and easily encorporated into the build.
+
+#### Engine Strategy
+
+Introducing `Strategy` as a concept allows us to re-think the way we approach
+Ember Engines as well. We can start thinking about engines as a strategy of
+certain type. In fact, it's very similar to `ConcatenationStrategy` described
+above.
+
++ reduces the API surface (as opposed to `ember-engines` and
+  `ember-asset-loader`, only one strategy)
++ makes things explicit, meaning that you have to register a strategy
+
+## Registering custom strategies
+
+Ember applications would be able to register custom strategies via
```

Comment:
What about Ember addons? I believe a lot of addons like fastboot may want to provide their own strategies that should be registered as well.